### ISSUE TITLE
profileManager.get(true) randomly throws java.util.concurrent.TimeoutException

### ISSUE THREAD

**Issue by vidma** [pac4j/play-pac4j#201]
Context: `play 2.6.3`,  `"org.pac4j" % "play-pac4j" % "4.0.0-SNAPSHOT"` and `pac4jVersion = "2.1.0"`.

I occasionally get `java.util.concurrent.TimeoutException` thrown by `profileManager.get(readFromSession = true)` (it's hanging/blocking for a few seconds where it shouldn't, and throws the exception afterwards). 

this code is called from a controller action (basically based on example in https://github.com/pac4j/play-pac4j-scala-demo/blob/0e2a7fa80e5e7cd177d72b90a707174f04eb4fe0/app/controllers/ApplicationWithFilter.scala#L22-L27 )
```scala
  def getProfile(implicit request: RequestHeader) = {
    import scala.collection.JavaConverters._
    val webContext = new PlayWebContext(request, playSessionStore)
    val profileManager = new ProfileManager[CommonProfile](webContext)
    val profile = profileManager.get(true) // randomly throws java.util.concurrent.TimeoutException
    profile
  }
```

I use:

```scala
// in SecurityModule.configure:
bind(classOf[PlaySessionStore]).to(classOf[PlayCacheSessionStore])

// and in build.sbt
libraryDependencies += ehcache
```

https://gist.github.com/vidma/a7c2154e15f74285c77b48508a78f066

**Comment by vidma** [comment]
cc @leleuj

**Comment by leleuj** [comment]
Do you have some specific configuration for EhCache? I mean: is it shared between nodes or something which could explain the hanging...

**Comment by millross** [comment]
Also any idea what your load is? The get could be doing blocking i/o under the hood, if distributed you might be blocking on a network call. Under load this would escalate very quickly if you have a limited threadpool for play to work on. I'm not sure if play assumes non-blocking i/o or not though, so this is conjecture.

**Comment by vidma** [comment]
it's default play cache config (i.e. no ehcache config).
I get the error on my laptop with `sbt run`, with only a single request at once...

**Comment by leleuj** [comment]
Have you already encountered this problem with the 4.0.0-SNAPSHOT or is it a very recent issue?

**Comment by vidma** [comment]
I tried 4.0.0-RC1 which never worked (was always throwing this error or a similar one), and `4.0.0-SNAPSHOT` works 4/5 times (since yesterday or the day before).

just in case FYI, here's an excerpt from my build.sbt:

```scala
scalaVersion := "2.12.2"
val playVersion = "2.6.3"
val pac4jVersion = "2.1.0"
// see https://github.com/pac4j/play-pac4j-scala-demo/blob/20ccf821bc557347ca2e555fb1c85d4afea92366/build.sbt#L9-L29
val pac4jSecurityDeps = Seq(
  "org.pac4j" % "play-pac4j" % "4.0.0-SNAPSHOT" excludeAll ExclusionRule("org.pac4j", "pac4j-core") changing(),
  "org.pac4j" % "pac4j"  % pac4jVersion changing(),
  "com.typesafe.play" %% "play-cache" % playVersion,
  "org.pac4j" % "pac4j-core" % pac4jVersion changing(),
  //   "org.pac4j" % "pac4j-sql" % "2.1.0",
  "org.pac4j" % "pac4j-http" % pac4jVersion changing(),
  // needed for auth session storage
  ehcache
)
libraryDependencies += ehcache
```

**Comment by vidma** [comment]
even more weird with play 2.6.5 - it logs out after a few visits.
weird enough, downgrading to `play 2.6.2` - seems to work so far. at least I didn't notice issues yet over ~20 visits.

does it have anything to do with how play-pac4j is built ?

**Comment by aleksandrmelnyk** [comment]
same issue here - play 2.6.5, after successful login, one request isAuthenticated true, all after that all next requests got false, session killed, no profiles retrieved from session. Easy to replicate, just need to upgrade play framework with basic/default dependencies and configuration.

**Comment by aleksandrmelnyk** [comment]
also can confirm: 
play 2.6.2 - seems working
play 2.6.3 - timeoutexception
play 2.6.5 - session killed after first request after authentication

**Comment by leleuj** [comment]
OK. Very strange that a minor upgrade of the Play framework breaks things so strongly. Will do some tests.

**Comment by leleuj** [comment]
Even stranger, I cannot reproduce the issue with both demos (Java and Scala). I updated them to Play 2.6.5 with success: @vidma @aleksandrmelnyk can you confirm you have the issue with at least one of the two demos?

**Comment by leleuj** [comment]
Demos: https://github.com/pac4j/play-pac4j-scala-demo and https://github.com/pac4j/play-pac4j-java-demo

**Comment by aleksandrmelnyk** [comment]
weird, but with the scala demo could not replicate on 2.6.5. Will debug later more.

**Comment by leleuj** [comment]
@aleksandrmelnyk Thanks. Weird indeed, but at least, it's a clue to find out the cause of the issue.

**Comment by millross** [comment]
Does it depend what session storage is being used?

**Comment by leleuj** [comment]
@millross I don't think so, as we all use EhCache (as defined in the dependencies)

**Comment by vidma** [comment]
@leleuj to replicate on 2.6.3 you need to visit the page which reads the profile from session (`profileManager.get(true)`, expecting timeoutexception ).

tried 2.6.5 and 2.6.3 demo, however unable to replicate so far (but had seen the issue on our app). I have no time to dig this deeper now.

---

 maybe it depends on how/where is the session accessed in the controller or is the anonymous client enabled (we had only FormClient, and used URL-based security filter rules). 

P.S. I also used 

```scala
@Singleton SomeController @Inject()(..., cc: ControllerComponents) 
  extends AbstractController(cc) with I18nSupport {
  ...
}
``` 

instead of more limiting `extends Security[CommonProfile]`.

**Comment by aleksandrmelnyk** [comment]
@leleuj Actually I just noticed that you didn't upgrade the scala demo project. Please, in project/plugins.sbt, change
`addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.6.2")`
to
`addSbtPlugin("com.typesafe.play" % "sbt-plugin" % "2.6.5")
`

**Comment by leleuj** [comment]
Good catch! I get the errors now. I will investigate further...

**Comment by leleuj** [comment]
With Play 2.6.3, I encounter the `TimeouException`, it seems that the problem comes from the Play cache implementation. I read several comments about that and found: https://github.com/playframework/playframework/pull/7802

@gmethvin is this really fixed in Play 2.6.5?

With Play 2.6.5, the authentication works but the profile does not seem to be kept in the cache more than a few milliseconds.

Has something changed with the timeout definition?

**Comment by leleuj** [comment]
OK. I debugged into the cache and indeed, the timeout has changed from seconds to milliseconds between 2.6.2 and 2.6.5. This explains the fact the profile is not saved long enough in the cache and is expired at the second call...

I will certainly change the default timeout in play-pac4j!
